src/BlockTypeRegistry.cpp

```diff
@@ -0,0 +1,153 @@
+
+#include "Globals.h"
+#include "BlockTypeRegistry.h"
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// BlockInfo:
+
+BlockInfo::BlockInfo(
+	const AString & aPluginName,
+	const AString & aBlockTypeName,
+	std::shared_ptr<cBlockHandler> aHandler,
+	const std::map<AString, AString> & aHints,
+	const std::map<AString, BlockInfo::HintCallback> & aHintCallbacks
+):
+	mPluginName(aPluginName),
+	mBlockTypeName(aBlockTypeName),
+	mHandler(aHandler),
+	mHints(aHints),
+	mHintCallbacks(aHintCallbacks)
+{
+}
+
+
+
+
+
+AString BlockInfo::hintValue(
+	const AString & aHintName,
+	const BlockState & aBlockState
+)
+{
+	// Search the hint callbacks first:
+	auto itrC = mHintCallbacks.find(aHintName);
+	if (itrC != mHintCallbacks.end())
+	{
+		// Hint callback found, use it:
+		return itrC->second(mBlockTypeName, aBlockState);
+	}
+
+	// Search the static hints:
+	auto itr = mHints.find(aHintName);
+	if (itr != mHints.end())
+	{
+		// Hint found, use it:
+		return itr->second;
+	}
+
+	// Nothing found, return empty string:
+	return AString();
+}
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// BlockTypeRegistry:
+
+void BlockTypeRegistry::registerBlockType(
+	const AString & aPluginName,
+	const AString & aBlockTypeName,
+	std::shared_ptr<cBlockHandler> aHandler,
+	const std::map<AString, AString> & aHints,
+	const std::map<AString, BlockInfo::HintCallback> & aHintCallbacks
```

Comment:
Perhaps we should make the registering of hints a different method. That way a plugin can extend existing functionality without having to completely override a previous block registration.

Reply:
Excellent idea, I'll add that.